src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2386,11 +2386,37 @@ ManageWorkerPool(WorkerPool *workerPool)
 			connectionFlags |= OUTSIDE_TRANSACTION;
 		}
 
+		if (CanUseOptionalConnection(workerPool))
```

Comment:
CanUse sounds a bit strange. How about ShouldWaitForConnection?

Reply:
> can't we always use an optional connection?

As far as I can say, no. The first connection of the execution is not optional.

> How about ShouldWaitForConnection?

That might be a little confusing as we're setting `OPTIONAL_CONNECTION ` flag -- given that we also have `WAIT_FOR_CONNECTION `.

Maybe `EligableForOptionalConnection()` ?

Reply:
I was thinking of inverting the meaning:
```c
		if (UseConnectionPerPlacement())
        {
			/*
		     * User wants one connection per placement, so no throttling is desired
             * and we do not set any flags.
             *
    		 * The primary reason for this is that allowing multiple backends to use
	    	 * connection per placement could lead to unresolved self deadlocks. In other
    		 * words, each backend may stuck waiting for other backends to get a slot
	    	 * in the shared connection counters.
			 */
        }
		else if (ShouldWaitForConnection(workerPool))
		{
			/*
			 * We need this connection to finish the execution. If it is not
             * available based on the current number of connections to the worker
             * then wait for it.
			 */
			connectionFlags |= WAIT_FOR_CONNECTION;
		}
		else
		{
			/*
			 * The executor can finish the execution with a single connection,
			 * remaining are optional. If the executor can get more connections,
			 * it can increase the parallelism.
			 */
			connectionFlags |= OPTIONAL_CONNECTION;
		}
```

with

```c
static bool
ShouldWaitForConnection(WorkerPool *workerPool)
{
	if (list_length(workerPool->sessionList) == 0)
	{
		/*
		 * We definitely need at least 1 connection to finish the execution.
		 * All single shard queries hit here with the default settings.
		 */
		return true;
	}

	if (list_length(workerPool->sessionList) < MaxCachedConnectionsPerWorker)
	{
		/*
		 * Until this session caches MaxCachedConnectionsPerWorker connections,
		 * this might lead some optional connections to be considered as non-optional
		 * when MaxCachedConnectionsPerWorker > 1.
		 *
		 * However, once the session caches MaxCachedConnectionsPerWorker (which is
		 * the second transaction executed in the session), Citus would utilize the
		 * cached connections as much as possible.
		 */
		return true;
	}

	return false;
}
```

Reply:
Makes sense, it is easier to read